### PR TITLE
Improve local scheduler API

### PR DESCRIPTION
**Pull request by skhalash**
**Description**

Changes proposed in this pull request:

- Split `LocalScheduler` and `RemoteScheduler` as well as the tests into separate files
- Improve `LocalScheduler` factory function by introducing options for optional params (`WithLogger`, `WithCRDComponents`, etc.)
- Add support for prioritized components (prereqs and components with CRDs)
- Simplify error handling for parallel reconciliation by using the `errgroup` package

**Note**
Even though there is some code duplication in `LocalScheduler` and `RemoteScheduler`, it was decided not to address it at the moment before CLI integration is complete.

**Related issue(s)**

### CODE REVIEW

File: pkg/scheduler/localscheduler.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+		return fmt.Errorf("failed to convert to cluster state: %s", err)
+	}
+
+	components, err := clusterState.Configuration.GetComponents()
+	if err != nil {
+		return fmt.Errorf("failed to get components: %s", err)
+	}
+
+	for _, c := range components {
+		if contains(ls.prerequisites, c.Component) {
```
Comment (lindnerby) [inline pkg/scheduler/localscheduler.go:70]: Suggestion: Reduce nesting by continuing early:
```suggestion
		if !contains(ls.prerequisites, c.Component) {
		     continue
		}
```

File: pkg/scheduler/localscheduler.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+			err = worker.Reconcile(c, *clusterState, schedulingID, false)
+			if err != nil {
+				return fmt.Errorf("failed to reconcile a component: %s", c.Component)
+			}
+		}
+	}
+
+	for _, c := range components {
+		if contains(ls.componentsWithCRDs, c.Component) {
```
Comment (lindnerby) [inline pkg/scheduler/localscheduler.go:84]: Same here, see above.

File: pkg/scheduler/localscheduler.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	return g.Wait()
+}
+
+func localClusterState(c *keb.Cluster) (*cluster.State, error) {
+	var defaultContractVersion int64 = 1
+
+	metadata, err := json.Marshal(c.Metadata)
+	if err != nil {
+		return nil, err
+	}
```
Comment (lindnerby) [inline pkg/scheduler/localscheduler.go:123]: Some empty lines after all if-statements of this function

**Review by lindnerby — COMMENTED** [review]
Just an idea: Maybe since many operations on collections here, we introduce some functions for working with component collections?
https://gobyexample.com/collection-functions
Filter for example?

File: pkg/scheduler/localscheduler.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+type LocalSchedulerOption func(*LocalScheduler)
+
+func WithLogger(logger *zap.SugaredLogger) LocalSchedulerOption {
+	return func(ls *LocalScheduler) {
+		ls.logger = logger
+	}
+}
+
+func WithComponentsWithCRDs(components []string) LocalSchedulerOption {
```
[the remark this answers is not in the document]
Comment (skhalash) [reply pkg/scheduler/localscheduler.go]: Agree, done

File: pkg/scheduler/localscheduler.go (hunk 1 above)
Comment (skhalash) [reply pkg/scheduler/localscheduler.go]: Done

File: pkg/scheduler/localscheduler.go (hunk 2 above)
Comment (skhalash) [reply pkg/scheduler/localscheduler.go]: Done

**Review by tobiscr — CHANGES_REQUESTED** [review]

File: cmd/local/cmd.go [hunk 5]
```diff
@@ -62,11 +62,11 @@ func RunLocal(o *Options) error {
 		},
 		true)
 
-	ls, _ := scheduler.NewLocalScheduler(keb.Cluster{
+	ls := scheduler.NewLocalScheduler(workerFactory, scheduler.WithLogger(l))
+	return ls.Run(context.Background(), keb.Cluster{
```
[the remark this answers is not in the document]
Comment (skhalash) [reply cmd/local/cmd.go]: Done

**Review by tobiscr — APPROVED** [review]

**Pull request merged**